skimage/morphology/tests/test_rolling_ball.py

```diff
@@ -0,0 +1,39 @@
+"""
+Tests for Rolling Ball Filter
+(skimage.morphology.rolling_ball)
+
+Author: Sebastian Wallkötter
+"""
+
+import numpy as np
+
+from skimage import data
+from skimage.morphology import rolling_ball
+from skimage.util import invert
+
+from skimage._shared.testing import assert_equal, fetch
+from skimage._shared import testing
+
+
+class TestRollingBall(object):
+    def test_white_background(self):
+        img = 23 * np.ones((100, 100), dtype=np.uint8)
+        img = invert(img)
+        result, bg = rolling_ball(img, radius=1, white_background=True)
+        assert np.allclose(result, 255 *np.ones_like(img))
+        assert np.allclose(bg, img)
+
+    def test_black_background(self):
+        img = 23 * np.ones((100, 100), dtype=np.uint8)
+        result, bg = rolling_ball(img, radius=1)
+        assert np.allclose(result, np.zeros_like(img))
+        assert np.allclose(bg, img)
```

Comment:
I'd love to see some more comprehensive tests, for example a test with a small image on which we can calculate the expected background. For example, an image whose brightness follows a sphere of a given radius *but* in which the central pixel is reduced by some value should have that value everywhere except a value of 0 at the central pixel. Correct?

Reply:
Yes, some more comprehensive tests can be useful. I didn't put much thought into this yet; mainly added this to get the ball rolling (pun intended).

The image you suggest will not be filtered in the way you describe. Assuming the specified `radius` is small enough to fit into the sphere, then the center pixel will indeed be 0; However, the remaining pixel intensities will be reduced by the "cap" of the ball that fits into the "gap" between the center pixel and the border, which can be calculated in advance, but it's not that trivial.

Could a slightly easier first step be to use images with linear intensity profile, e.g., I(x,y) = 10*x + 5*y + 6? This should result in a filtered image with 0 intensity everywhere.

Reply:
> mainly added this to get the ball rolling (pun intended).

😂 

> the remaining pixel intensities will be reduced by the "cap" of the ball that fits into the "gap" between the center pixel and the border, which can be calculated in advance, but it's not that trivial.

Of course! I forgot to roll my ball 😂.

Ok, so one way to make this a bit easier is to make a 1-pixel wide, super skinny image. Then you can think about a circle instead of a ball, which I at least find much easier, and calculate the height of the midpoint of various dips. Does that make sense?

> Could a slightly easier first step be to use images with linear intensity profile, e.g., I(x,y) = 10x + 5y + 6? This should result in a filtered image with 0 intensity everywhere.

This is only slightly better than constant intensity...

Another idea: image with just a single small radius sphere as signal, and a larger radius sphere used for background removal. Then the intensity at the centre of the small sphere should be reduced by, again, a nontrivial amount, but one that should be reasonably easy to calculate for given radii.

Reply:
I've added some more tests in firefoxmetzger/scikit-image@36e3c827434e9b9297361e018715d9e185b7834d. Nothing super complex, but more than just testing a constant image. I am now testing

- a constant image for multiple radii
- a radial gradient on the image (assumed to be uneven exposure)
- a linear gradient on the image (assumed to be uneven exposure)
- 1 pixel wide features on an image with a constant background for multiple radii

Reply:
How about moving the spheroid profile calculation that creates the kernel into a separate function? That way you could test that function against the background subtraction. Something like this:
```python
def shperoid_kernel(kernel_size=(100, 100), intensity_vertex=(100,))
    kernel_size = np.array(kernel_size)
    if not np.issubdtype(kernel_size.dtype, np.number):
        raise ValueError(
            "kernel_size must be convertible to a numeric array.")
    if not kernel_size.size == 2 and len(kernel_size.shape) == 2:
        raise ValueError(
            "kernel_size must be a one dimensional array_like of size 2.")
    if np.any(kernel_size <= 0):
        raise ValueError(f"All elements of kernel_size must be greater zero.")
    kernel_size = kernel_size / 2

    intensity_vertex = np.array(intensity_vertex)
    if not np.issubdtype(intensity_vertex.dtype, np.number):
        raise ValueError(
            "Intensity_vertex must be convertible to a numeric array.")
    if not intensity_vertex.size == 1 and len(intensity_vertex.shape) == 1:
        raise ValueError(
            "Intensity_vertex must be a scalar.")
    if np.any(intensity_vertex <= 0):
        raise ValueError(f"Intensity_vertex must be greater zero.")
    intensity_vertex = intensity_vertex / 2

    kernel_size_y, kernel_size_x = np.round(kernel_size).astype(int)

    tmp_x = np.arange(-kernel_size_x, kernel_size_x + 1)
    tmp_y = np.arange(-kernel_size_y, kernel_size_y + 1)
    x, y = np.meshgrid(tmp_x, tmp_y)

    kernel_size_y, kernel_size_x = kernel_size
    tmp = (x / kernel_size_x) ** 2 + (y / kernel_size_y) ** 2
    cap_height = intensity_vertex - intensity_vertex * \
        np.sqrt(np.clip(1 - tmp, 0, None))

    kernel = np.array(tmp <= 1, dtype=float)
    kernel[kernel == 0] = np.Inf
    return kernel


def test_spheroid_gradient():
    # spheroid image
    img = spheroid_kernel(kernel_size=(100, 100), intensity_vertex=(100,))

    result = rolling_ball(img, radius=5)
    assert np.allclose(result, np.zeros_like(img))

def test_small_spheroid_gradient():
    # spheroid image
    img = spheroid_kernel(kernel_size=(5, 5), intensity_vertex=(5,)) # this feature may have to be embedded into a larger image

    result = rolling_ball(img, radius=100)
    assert not np.allclose(result, np.zeros_like(img))
```
This should definitely only be in addition to the other tests, but it is a nice way to confirm that the background subtraction works as intended